XProc code refactoring and deduplication, bugfixes, documentation, and tests

bertfrees opened the pull request:
@rdeltour I know this looks like a lot to digest (and probably is), but a big part of it is refactorings that can be skimmed through.

This is a summary of the changes:

- Wrap XSLTs in public steps when appropriate
  - `px:epub3-nav-to-ncx`
  - `px:odf-manifest-from-fileset`
  - `px:html-id-fixer`
  - `px:smil-to-audio-fileset`
  - `px:smil-to-text-fileset`
  - `px:html-upgrade`
  - `px:html-downgrade`
  - `px:html-fixer`
  - `pef:add-metadata`
  - `px:dtbook-to-opf-metadata`
  - ...
- Improvements applied to many steps
  - **Give various steps "fileset" and "in-memory" inputs and outputs**
  - Support resources on "in-memory" ports of steps
  - **Make steps load files that are not in memory yet**
  - Make various steps set the base URI of the result document
  - **Avoid setting `xml:base` attribute (setting the base URI can be done without it)**
- Improvements to utility libraries - related to filesets
  - Perform URI normalization before comparing URIs in various `fileset-utils` steps
  - Create more consistency in when "original-href" attributes are present in filesets and what it means
  - Change behavior of `px:fileset-copy` and `px:fileset-move` so that they don't make any changes on disk (leave this up to `px:fileset-store` and `px:fileset-delete`)
  - Add "fileset" input and/or output ports to various fileset-utils steps (`px:fileset-load`, `px:fileset-store`)
  - Add "in-memory" input and/or output ports to various fileset-utils steps (`px:fileset-add-entry`, `px:fileset-filter`, `px:fileset-filter-in-memory`)
  - Make "in-memory" input optional on `px:fileset-load` and `px:fileset-store`
  - `px:fileset-add-entry`: allow specifying custom file attibutes via parameters
  - **`px:fileset-add-entry`: add a "replace" option**
  - `px:fileset-filter`: add a "not-matched" output port
  - `px:fileset-filter`: allow media-types to be separated by more than just a single space
  - `px:fileset-filter-in-memory`: add a "not-in-memory" output port
  - `px:fileset-load`: handle special characters in zip URIs
  - `px:fileset-load`: support loading text or HTML files from ZIP
  - `px:fileset-load`: replace "method" option with support for "method" attributes
  - `px:fileset-load`: load files with media type "text/xml" as XML
  - `px:fileset-join`: make the behavior more predictable
  - `px:fileset-store`: make sure that serialization attributes to not affect files that do not come from memory
  - `px:fileset-store`: make "doc-type" attributes work also when storing to ZIP
  - `px:fileset-store`: support "xml-declaration" attributes
  - `px:fileset-store`: support storing text files to ZIP
  - `px:fileset-store`: optimize for files that originate from a ZIP
  - New `px:fileset-filter-in-memory` step (was private)
  - New `px:fileset-delete` step
  - New `px:fileset-purge` step
  - **New `px:fileset-apply` step to apply rename actions defined in a "mapping" document**
  - **New `px:fileset-compose` step to compose two mapping documents**
  - **New `px:fileset-invert` to create the inverse of a mapping document**
  - Add "mapping" output to `px:fileset-copy`
  - **Add "mapping" output to `px:html-add-ids`**
  - **Add "mapping" output to `px:epub3-safe-uris`**
  - **Add "mapping" output to `px:diagram-to-html`**
  - **Add "mapping" output to `px:chunker`**
  - **Add "mapping" output to `px:dtbook-merge`, `px:dtbook-to-zedai`, `px:zedai-to-html` and `px:dtbook-to-html`**
- Improvements to other utility libraries
  - **New `px:log-error` step to create a message in the job execution log from an XProc error and to print the stack trace in the detailed log**
  - **`px:error`: don't add leading and trailing white space to message**
  - **Add `pf:numeric-decimal-to-alpha` to numeral-conversion.xsl**
  - Make "href" option optional on `px:tempfile` and `px:tempdir`
  - `px:tempdir`: add "delete-on-exit" option
  - `pf:relativize-uri`: fix for MacOS
  - `pf:expand83`: support zip URIs
  - New steps `px:read-doctype` and `px:read-xml-declaration`
  - **New custom XProcSpec assertion to check the XML declaration of a file**
  - Perform `px:mediatype-detect` within `px:html-to-fileset`
  - **Make `px:html-add-ids` more robust: handle elements with `epub:type="pagebreak foo"`**
  - **Make `px:html-add-ids` configurable: add a "match" option to specify the elements that need an ID**
  - Simplify `px:epub3-create-mediaoverlays` interface
  - Simplify `px:epub3-pub-create-package-doc` interface
  - **New `px:epub3-pub-add-mediaoverlays` step**
  - New `px:epub3-pub-merge-prefix`step
  - New `px:epub3-pub-add-prefix`step
  - New `px:epub3-safe-uris` step
  - **New `px:epub3-ensure-core-media` step to ensure that a EPUB publication contains only resources that are EPUB 3 core media types**
  - **New `px:epub3-nav-from-ncx` step**
  - **New `px:epub3-validate` step**
  - New `epub3-vocab.xsl` in `epub3-utils` with utility functions related to prefix attribute
  - `px:epub3-ocf-finalize`: set `media-type="application/xml"` on generated OCF files
  - `px:epub3-pub-create-package-doc`: add a "reserved-prefixes" option to control which prefixes are implicit in the resulting package document (declare prefixes systematically by default)
  - `px:epub3-pub-create-package-doc`: let the spine order be determined by the `source.fileset` port
  - `px:epub3-pub-create-package-doc`: fix "bindings" input
  - **`px:epub3-create-mediaoverlays`: optimization: skip if no audio**
  - **Make `px:epub3-safe-uris` more robust: make sure result does not have duplicate file names**
  - **`px:epub-load`: add compression-method "stored" to mimetype and put it as first file**
  - **`px:epub-load`: add "validation" option**
  - **Simplify `px:daisy3-create-smils`, `px:daisy3-create-ncx` and `px:daisy3-create-opf` interfaces**
  - **Simplify `px:tts-for-epub3` and `px:tts-for-dtbook` interfaces**
  - **Set media-type and doctype on result filesets (`px:daisy3-create-smils`, `px:daisy3-create-ncx`, `px:daisy3-create-res-file`, `px:daisy3-create-opf`, `px:daisy3-prepare-dtbook`)**
  - **Add "fail-if-missing-smilref" option to `px:daisy3-create-ncx`**
  - **New `px:daisy3-add-smilrefs` step to add smilref attributes to a DTBook based on a set of SMIL documents**
  - **New `px:dtbook-validate` step**
  - **Generalize `px:mediaoverlay-upgrade-smil` to `px:smil-upgrade`**
  - **New generic `px:smil-to-audio-clips` step**
  - **New `px:smil-downgrade` step**
  - **`px:odf-store`: handle case when fileset already contains mimetype**
  - **text-to-ssml: support loading PLS from ZIP**
  - `px:ssml-to-audio` (and steps that make use of it): change "output-dir" option into "log" port
  - **New `px:html-update-links`, `px:epub-update-links`, `px:daisy202-update-links`, `px:dtbook-update-links`, `px:smil-update-links` and `px:ncx-update-links` steps to update cross-references after resources have been relocated and/or id attributes modified**
  - **New `px:epub3-rename-files` and `px:daisy202-rename-files` steps to rename files in a EPUB/DAISY 2.02 fileset (and update cross-references)**
  - **Make generated id attributes more predictable (to fix brittle tests) in `px:break-and-reshape`, `px:smil-downgrade`, `px:dtbook-to-zedai`, `px:zedai-to-opf-metadata`, ...**
  - ...
- Merge/split/rename modules and move files around
  - Merge `epub3-nav-utils`, `epub3-pub-utils` and `epub3-ocf-utils` into `epub-utils` (**also drop the division in sub-libraries "pub", "nav" and "ocf"**)
  - Merge `lexing` and `break-detection` modules into `nlp-common`
  - Merge module `tts-helpers` into `tts-common`
  - Merge `epub3-to-ssml` module into `epub3-tts`
  - Merge `dtbook-to-ssml` module into `dtbook-tts`
  - Rename `odt-utils` to `odf-utils`
  - Rename `mediaoverlay-utils` to `smil-utils`
   - Move MathML DTD from `common-entities` to new `mathml-utils` module
  - Move MathML RelaxNG schema from ` dtbook-utils` to `mathml-utils`
  - Move ZedAI schema from `dtbook-to-zedai` to `zedai-utils` (new `px:zedai-validate` step)
  - Remove some DTD from `daisy202-utils` that were also in `html-utils`
  - Move schema files from `daisy202-validator` to `daisy202-utils`
  - Remove SMIL DTD from `common-entities` and make the ones in `smil-utils` public
  - Move XHTML DTD's from `common-entities` to `html-utils`
  - Move NCX and OEB DTD's from `common-entities` to `daisy3-utils`
  - Move DTBook DTD's from `common-entities` to dtbook-utils
  - Delete `common-entities`
  - Move `px:html-downgrade` from `epub3-to-daisy202` to `html-utils`
  - Move `dtbook-to-metadata.xsl` from `dtbook-to-pef` to `dtbook-to-epub3`
  - Move `dtbook-annotating.xsl` from `text-to-ssml` to `dtbook-tts`
  - Move `pxi:convert-diagram-descriptions` from `html-to-epub3` to `px:diagram-to-html` in `zedai-to-html`
  - Move `px:create-mediaoverlays` from `mediaoverlay-utils` to `px:epub3-create-mediaoverlays` in `epub-utils`
  - Move `px:opf-manifest-to-fileset` from `epub3-utils/ocf/library.xpl` to `epub3-utils/pub/library.xpl`
  - Move `px:opf-spine-to-fileset` from `epub3-utils/ocf/library.xpl` to `epub3-utils/pub/library.xpl`
  - **Generalize `px:epub3-load` to `px:epub-load` and `px:epub3-validate` to `px:epub-validate`**
  - Replace `px:create-audio-files` and `px:rm-audio-files` in `tts-helpers` with `px:audio-clips-to-fileset` and `px:audio-clips-update-files` (new in `smil-utils`) and `px:fileset-move` and `px:fileset-delete`
  - **Make certain steps private (`px:html-load`, `px:epub3-ocf-zip`, `px:opf-manifest-to-fileset`, `px:epub3-nav-aggregate`)**
  - **Rename `px:inline-css-speech` to `px:css-speech-cascade` (for symmetry with `px:css-cascade`)**
  - **Rename `px:remove-inline-css-speech` to `px:css-speech-clean`**
  - **Rename `px:create-ncx` to `px:daisy3-create-ncx`**
  - **Rename `px:create-daisy3-opf` to `px:daisy3-create-opf`**
  - **Rename `px:create-res-file` to `px:daisy3-create-res-file`**
  - **Rename `px:create-daisy3-smils` to `px:daisy3-create-smils`**
  - **Rename `px:isolate-daisy3-skippable` to `px:daisy3-isolate-skippable`**
  - **Rename `px:daisy3-fix-dtbook-structure` to `px:daisy3-prepare-dtbook`**
  - **Rename `px:epub3-nav-to-guide` to `px:epub-landmarks-to-guide`**
  - **Rename `px:merge-dtbook` to `px:dtbook-merge`**
  - **Rename `px:upgrade-dtbook` to `px:dtbook-upgrade`**
  - **Rename `px:html-id-fixer` to `px:html-add-ids`**
  - Major refactoring of `css-utils` (the main one and the braille one). Notably `css:inline` has been moved to the main `css-utils` and has been renamed to `px:css-cascade`.
  - ...
- Move code around, remove duplication, make more use of other steps/modules, ...
  - Make more use of the various `fileset-utils` steps
  - **Eliminate some ".load" and ".store" steps**
  - Use `mathml-utils` in `dtbook-utils`
  - Use `mathml-utils` in `zedai-utils`
  - Use `px:odf-manifest-from-fileset` in `px:epub3-ocf-finalize`
  - Use `px:zedai-validate` in `dtbook-to-zedai`
  - **Create new `px:epub3-add-navigation-doc` that combines `px:epub3-nav-create-toc`, `px:epub3-nav-create-page-list` and `px:epub3-nav-aggregate`, and use it in `px:epub3-pub-create-package-doc`, `px:zedai-to-epub3` and `px:html-to-epub3`**
  - Use `px:odf-store` in `px:epub3-ocf-zip`
  - Use `px:dtbook-to-html` in `px:daisy3-to-epub3`
  - Use `px:zedai-to-html` in `zedai-to-epub3`
  - Use `px:epub3-safe-uris` in `px:html-to-epub3`
  - Use `px:html-to-epub3` in `px:zedai-to-epub3`
  - Use improved `px:html-to-opf-metadata` from `html-to-epub3` in `px:html-to-pef`
  - Use `px:data` in `pxi:load-html`
  - **Use mapping output of `px:epub3-safe-uris` in `html-to-epub3`**
  - **Use `px:html-update-links`, `px:smil-update-links` and `px:fileset-apply` in `px:epub3-safe-uris`**
  - **Use `px:html-update-links` in `px:diagram-to-html`**
  - **Use `px:html-update-links` in daisy3-to-epub3 (ncx-to-nav)**
  - **Use `px:epub-update-links` in `px:epub3-to-epub3 `**
  - **Use `px:daisy202-rename-files` in `px:epub3-to-daisy202`**
  - **Use `px:css-speech-cascade` in `px:tts-for-epub3` and `px:tts-for-dtbook`**
  - **Perform `px:html-add-ids` inside `px:html-outline` and `px:epub3-create-page-list`**
  - **Use "fix-untitled-sections-in-outline" option of `px:html-outline` in `px:epub3-create-toc.`**
  - **Use `px:html-outline` in `px:html-downgrade`**
  - **`px:html-chunker`: use `px:html-update-links` and "mapping" output of `px:chunker`, and remove "link-attribute-name" option**
  - **Use `px:epub3-nav-from-ncx` in daisy3-to-epub3**
  - **Extract generic steps `px:xml-chunker` and `px:add-ids` from html-utils into common-utils**
  - **Use `px:xml-chunker` in `css:split`**
  - **Add the MathML altimg fallback within `px:daisy3-prepare-dtbook`**
  - ...
- Rename step inputs/outputs/options to make the naming more consistent across steps
- Restructure code to make it simpler and more readable
- **Rewrite code in XSLT or XProc**
- **Change "tts-config" script options to input ports**
- Reformat code
- Add step documentation, comments, etc.
- Add XProcspec tests
- **Port XSpec tests to XProcSpec**
- Fix brittle tests
- **Add progress info to various steps**
- **Various bugfixes**
- ...

bertfrees commented on common/fileset-utils/src/main/resources/xml/xproc/fileset-add-entry.xpl:
```diff
@@ -1,11 +1,55 @@
+  </p:documentation>
+
+  <p:input port="source" primary="true">
+    <!-- FIXME: rename to source.fileset -->
```
Oops didn't intend to include this comment in the commit. You're right, because we're updating the major version number we could do it now. But I was thinking about doing a global rename of the ports anyway so we could also cover it then. The major version update ([Change behavior of px:fileset-copy and px:fileset-move](3453309b558b42bd15ad7fea989eafebb4fcac15)) was done after this change ([Add in-memory ports to px:fileset-add-entry](494500cb0e574e4873dd3873d2ee6d9a919450f0)) so I probably didn't even realize it could be done now.

bertfrees commented:
@rdeltour I have added a whole bunch of commits to the branch: https://github.com/daisy/pipeline-modules/compare/ee04748a4eaeb7b7c8b9ee0cc5355009f6b964df...51f120c80ca725f642c39c0497e2afe61f4ff444. I have updated the summary of the changes (new things are marked in bold).

rdeltour commented on scripts-utils/epub3-utils/src/main/resources/xml/epub3-vocab.xsl:
```diff
@@ -0,0 +1,286 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="2.0"
```
it might be useful to add a "cleanup" function that tries to clean and fix a prefix declaration that contain validation errors into something valid?

bertfrees replied:
Hmm OK but what kind of errors would it detect?

rdeltour replied:
merging duplicate prefix, replacing redeclared reserved prefixes with default ones, removing prefixed default vocabs?

bertfrees replied:
OK, so you mean more like a function or template on the whole document rather than a function that parses a single prefix attribute? That could be useful yes. I'm definitely going to add more stuff to this library, notably there is some NLB code that I would like to integrate (or replace) and that does things similar to what you are proposing.

bertfrees replied:
You should now be able to do this kind of cleanup with the new [`px:epub3-pub-add-prefix`](https://github.com/daisy/pipeline-modules/commit/82adb09dd9bc8837be3b4cdfe8d65c34117b81d9) or [`px:epub3-pub-merge-prefix`](https://github.com/daisy/pipeline-modules/commit/f0b6caade7973071b94e5b0ca8f17af9cd04fde9) step.

rdeltour commented on scripts-utils/daisy202-utils/src/main/resources/xml/xproc/load/load.xpl:
```diff
@@ -31,11 +31,40 @@
-    <p:import href="http://www.daisy.org/pipeline/modules/mediatype-utils/library.xpl"/>
+    <p:import href="http://www.daisy.org/pipeline/modules/common-utils/library.xpl">
+        <p:documentation>
+            px:message
```
out of curiosity, is it used for internal documentation only or for something more advanced?
Looks like a useful piece of info, although can be hard to maintain over time (we need to think about having to update this when the code is changed).

bertfrees replied:
It's just internal documentation. Yes it needs updating, but on the other hand it also makes it easier to detect when an import can be removed.

rdeltour replied:
ok good!

bertfrees commented on scripts-utils/epub3-utils/src/main/resources/xml/epub3-vocab.xsl:
```diff
@@ -0,0 +1,286 @@
+            </xsl:matching-substring>
+            <xsl:non-matching-substring>
+                <xsl:if test="not(matches(.,'[\s\t\n\r]+'))">
+                    <xsl:message terminate="yes" select="concat('Error parsing prefix attribute: ',$prefix-decl)"/>
```
I think I meant this to be more like a code assertion. I don't think the input will ever directly come from the user. I'll check to be sure. For assertions do you also think the error should become a warning?

rdeltour replied:
> For assertions do you also think the error should become a warning?
I don't know. Basically I think it depends on how this is called. As always our scripts need to be permissive in what they accept and produce valid output, so whatever we can fix we should, and we shouldn't reject an entire set of prefix declarations if only one of its declarations is incorrect.

bertfrees replied:
I meant that it is just a way to ensure that there are no bugs in other parts of the code. If the input comes directly from the user I think it's clear that the function needs to be more permissive than it is now, but what if the input comes from other parts of the code? Maybe we need something like assertions in Java which can be disabled?

rdeltour replied:
yes, I agree it largely depends on who calls this function.

bertfrees replied:
See commit https://github.com/daisy/pipeline-modules/commit/8dadc115d61d638c804f20e7a3c088b51185c387

bertfrees commented:
> Let me know if you need me to review specific files or scripts!

Well I was hoping you could look at everything, even if it's just a cursory look. I can't really predict which specific files or commits you will have remarks on. Probably the stuff under "Improvements to utility libraries" deserves a closer look than the other stuff, which falls under the "refactorings" category. It's a huge PR, but I've made a lot of small and independent commits, so you should be able to look at them one by one.

Before merging this PR I'm first going to add a few more improvements. For example, the "epub3-vocab.xsl " file that you had some remarks on will be further improved.

Also there is a remaining issue that we need to fix first. Running this code will fail because of XProc errors. The errors have to do with automatic port connections. I had to make some local adjustments to the XProc code to work around these errors, but I don't see anything wrong with the code. I suspect it's because the XProc engine is buggy.

bertfrees commented:
@rdeltour I've added some more commits to the branch: https://github.com/daisy/pipeline-modules/compare/f07db48be1f521bb14de6ca53c407d7358db7a12...1b26c844f2bed91f734e4cd62e4e944c7fbd6657. I've updated the summary again (new things again marked in bold)

> I suspect it's because the XProc engine is buggy.

I think I have found the bug. See https://github.com/daisy/xmlcalabash1/pull/3.

bertfrees commented:
Added commits https://github.com/daisy/pipeline-modules/compare/1b26c844f2bed91f734e4cd62e4e944c7fbd6657...8dadc115d61d638c804f20e7a3c088b51185c387

bertfrees commented on scripts-utils/epub3-utils/src/main/resources/xml/epub3-vocab.xsl:
```diff
@@ -0,0 +1,286 @@
+    <xsl:variable name="f:default-prefixes" as="element(f:vocab)*">
+        <f:vocab prefix="a11y"      uri="http://www.idpf.org/epub/vocab/package/a11y/#"/>
+        <f:vocab prefix="dcterms"   uri="http://purl.org/dc/terms/"/>
+        <f:vocab prefix="epubsc"    uri="http://idpf.org/epub/vocab/sc/#"/>
```
Fixed by commit https://github.com/daisy/pipeline-modules/commit/f454cedcb331d65e6010d44fb14562ddcc9a45b4

rdeltour commented on common/fileset-utils/src/main/resources/xml/xproc/fileset-load.xpl:
```diff
@@ -292,7 +292,7 @@
 
                   <!-- XML -->
-                  <p:when test="$media-type='application/xml' or matches($media-type,'.*\+xml$')">
+                  <p:when test="matches($media-type,'.*/xml$') or matches($media-type,'.*\+xml$')">
```
the expression could be slightly simplified with one `matches` using the `|` regex operator (but it's a matter of taste ;-)

bertfrees replied:
Missed that.

bertfrees replied:
Fixed

bertfrees commented on common/file-utils/src/main/resources/xml/xproc/set-base-uri.xpl:
```diff
@@ -4,6 +4,17 @@
+		an xml:base attribute. Elements with a relative xml:base are also affected by the
+		px:set-base-uri in that their absolute base URI changes accordingly, but the xml:base
+		attribute is not changed. Elements with an absolute xml:base, and their descendants, are not
+		affected by the px:set-base-uri.</p>
```
A link to the source code of the XProcSpec test on GitHub?

bertfrees commented:
> I maybe noted some small semver inconsistencies, but this needs to be double-checked by looking at the whole diff, not at the individual-commit level.

I did go through all commits and fixed wrong versions, but it could be that I missed some.

bertfrees commented on bom/pom.xml:
```diff
@@ -382,7 +382,7 @@
         <groupId>org.daisy.pipeline.modules</groupId>
         <artifactId>zedai-to-epub3</artifactId>
-        <version>2.1.0</version>
+        <version>2.1.1-SNAPSHOT</version>
```
Not breaking because I just added a port, so updating minor version (to 2.2.0) should be enough right?

bertfrees replied:
No, I looked wrong. Actually there is no change to the API at all in 6c9d98c5063bf4336c2eaa472a7898988742b68f.

rdeltour replied:
Sorry, I was looking at the other changes in this commit, which is about an API break in `epub3-nav-utils` (renaming a step option), but didn't realize this line above is not the module in question but is `zedai-to-epub3`.

bertfrees commented on bom/pom.xml:
```diff
@@ -287,7 +287,7 @@
         <groupId>org.daisy.pipeline.modules</groupId>
         <artifactId>daisy202-to-epub3</artifactId>
-        <version>2.0.1</version>
+        <version>2.0.2-SNAPSHOT</version>
```
There's no new option afaics.

bertfrees commented on scripts-utils/epub3-nav-utils/src/main/resources/META-INF/catalog.xml:
```diff
@@ -2,8 +2,6 @@
 <catalog xmlns="urn:oasis:names:tc:entity:xmlns:xml:catalog">
    <uri name="http://www.daisy.org/pipeline/modules/epub3-nav-utils/library.xpl"
         uri="../xml/xproc/epub3-nav-library.xpl"/>
-   <uri name="http://www.daisy.org/pipeline/modules/epub3-nav-utils/nav-to-ncx.xsl"
```
Yes I think this is a typical example of something that should be wrapped in a step. If it ever turns out we need to use it from another XSLT we can put it back in the API. I've done this in several places.

True about the version.

bertfrees commented on scripts-utils/epub3-utils/pom.xml:
```diff
@@ -55,6 +55,11 @@
+      <groupId>org.daisy.pipeline.modules</groupId>
+      <artifactId>odf-utils</artifactId>
+      <scope>runtime</scope>
+    </dependency>
```
I did this also because OCF is based on ODF. Even if you choose to skip the manifest, epub3-utils uses odf-utils for the zipping and the mimetype.

bertfrees commented on scripts-utils/odf-utils/src/main/resources/xml/manifest-from-fileset.xpl:
```diff
@@ -26,11 +27,28 @@
+		</p:documentation>
+	</p:import>
+	
+	<p:variable name="base2"
```
I know. I've struggled a bit with this. It should be called just "base" really. It's annoying that XProc doesn't allow variable shadowing like XSLT does.

bertfrees commented on common/html-utils/src/main/resources/xml/xslt/html-id-fixer.xsl:
```diff
@@ -49,9 +46,8 @@
-
     <xsl:template match="*[@epub:type='pagebreak']">
-        <!--TODO FIXME: epub:type can have several values-->
+        <!-- FIXME: epub:type can have several values -->
```
> maybe it could be fixed altogether

Yeah OK. Indeed not very hard to fix. But I didn't want to get too side-tracked. The amount of side-tracking was already enormous :-), but most other changes to the behavior that I did were needed for the dtbook-to-epub3 rewrite.

> ultimately this code would need to be adapted to also support DPUB ARIA roles

Can you create a new issue for this? Or do we have one for it?

rdeltour replied:
> Can you create a new issue for this?

Done in daisy/pipeline-tasks#140

bertfrees commented on tts/epub3-tts/src/main/resources/xml/tts-for-epub3.xpl:
```diff
@@ -1,9 +1,9 @@
-		xmlns:px="http://www.daisy.org/ns/pipeline/xproc"
-		xmlns:d="http://www.daisy.org/ns/pipeline/data"
-		exclude-inline-prefixes="#all">
+<p:declare-step xmlns:p="http://www.w3.org/ns/xproc" version="1.0"
```
Yes, although it sounds pretty challenging. Especially formatting multi-line attributes sounds hard to do.

bertfrees commented on scripts-utils/epub3-utils/src/main/resources/xml/pub/create-package-doc.fileset-to-manifest.xsl:
```diff
@@ -1,39 +1,68 @@
+                    <xsl:variable name="svg"
+                                  select="if (@svg='true') then concat($scripted,' svg') else $scripted"/>
+                    <xsl:variable name="switch"
+                                  select="if (@switch='true') then concat($svg,' switch') else $svg"/>
```
Yep good point.

bertfrees commented on scripts-utils/epub3-utils/src/main/resources/xml/epub3-safe-uris.xsl:
```diff
@@ -0,0 +1,50 @@
+                                escape-html-uri(
+                                  replace(
+                                    pf:unescape-uri(pf:get-path($uri)),
+                                    '[^\p{L}\p{N}\-/_.]','_')))"/>
```
True. This is not handled anywhere. I guess it's not that hard to fix in epub3-safe-uris.xsl.

bertfrees commented on scripts/html-to-epub3/src/main/resources/xml/xslt/html-to-metadata.xsl:
```diff
@@ -67,39 +70,48 @@
+                content="{format-dateTime(
+                    adjust-dateTime-to-timezone(current-dateTime(),xs:dayTimeDuration('PT0H')),
+                    '[Y0001]-[M01]-[D01]T[H01]:[m01]:[s01][Z]')}"
+            />
```
Hmm, I took this from nordic-dtbook-epub3-migrator without much thinking.
px:epub3-pub-create-package-doc also does this if the `compatibility-mode` option is set (see create-package-doc.backwards-compatible-metadata.xsl). Should we make it an option in html-to-epub3 too?

bertfrees replied:
Actually we can just remove the second `meta` element, and it will be created in `px:epub3-pub-create-package-doc`.

bertfrees commented on scripts-utils/epub3-utils/src/main/resources/xml/pub/create-package-doc.xpl:
```diff
@@ -22,6 +22,12 @@
             href="http://www.idpf.org/epub/301/spec/epub-publications.html#gloss-media-overlay-document">media
             overlay documents</a> and other resources.</p>
+            <p>If this fileset includes a navigation document, it should be marked with a
+            <code>nav</code> attribute with value <code>true</code>, and this file should be a
```
That would indeed be nicer.

bertfrees commented on common/fileset-utils/src/main/resources/xml/xproc/fileset-filter.xpl:
```diff
@@ -73,7 +73,7 @@
             <p:delete>
                 <p:with-option name="match"
-                    select="concat(&quot;//d:file[@media-type='' or not(some $media-type-regex in tokenize('&quot;,$media-types-regexes,&quot;',' ') satisfies matches(@media-type,$media-type-regex))]&quot;)"
+                    select="concat(&quot;//d:file[@media-type='' or not(some $media-type-regex in tokenize('&quot;,$media-types-regexes,&quot;','[\s\t\n]+')[not(.='')] satisfies matches(@media-type,$media-type-regex))]&quot;)"
```
Right, forgot about that.

bertfrees replied:
Fixed

bertfrees commented on bom/pom.xml:
```diff
@@ -150,7 +150,7 @@
         <groupId>org.daisy.pipeline.modules</groupId>
         <artifactId>mediatype-utils</artifactId>
-        <version>1.2.0</version>
+        <version>1.2.1-SNAPSHOT</version>
```
Why? I don't see any behavioral changes, just refactoring.

rdeltour replied:
Sorry false alarm, I hadn't realized the API change was about `fileset-utils` and the version above was `mediatype-utils`

bertfrees commented on scripts-utils/epub3-utils/src/main/resources/xml/pub/create-package-doc.xpl:
```diff
@@ -28,6 +28,10 @@
             document.</p>
+            <p>If the fileset includes a <a
+            href="http://www.idpf.org/epub/301/spec/epub-publications.html#cover-image"><code>cover-image</code></a>,
+            it should be marked with a <code>cover-image</code> attribute with value
```
Done

bertfrees commented on common/fileset-utils/src/main/resources/xml/xproc/fileset-filter.xpl:
```diff
@@ -87,7 +87,7 @@
             <p:delete>
                 <p:with-option name="match"
-                    select="concat(&quot;//d:file[not(@media-type='') and (some $not-media-type-regex in tokenize('&quot;,$not-media-types-regexes,&quot;',' ') satisfies matches(@media-type,$not-media-type-regex))]&quot;)"
+                    select="concat(&quot;//d:file[not(@media-type='') and (some $not-media-type-regex in tokenize('&quot;,$not-media-types-regexes,&quot;','[\s\t\n]+')[not(.='')] satisfies matches(@media-type,$not-media-type-regex))]&quot;)"
```
Fixed

bertfrees commented:
Added commits https://github.com/daisy/pipeline-modules/compare/0b53e307506ffb810c50ae8b46c957be619d300e...d6647ee and updated the summary (new stuff in bold).
